File: content/tracks/dsa-python/concepts/13-graphs-bfs-dfs/graphs-bfs-dfs_test.py
import sys
from stub import num_islands, can_finish, word_ladder_length

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


def grid(rows: list[list[str]]) -> list[list[str]]:
    return [row[:] for row in rows]


# num_islands
r = num_islands(grid([["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]))
test("islands: one large island", r == 1, f"got {r}")
r = num_islands(grid([["1","1","0","0","0"],["1","1","0","0","0"],["0","0","1","0","0"],["0","0","0","1","1"]]))
test("islands: three islands", r == 3, f"got {r}")
r = num_islands(grid([["0","0","0"],["0","0","0"]]))
test("islands: all water", r == 0, f"got {r}")
r = num_islands(grid([["1","0","1"],["0","1","0"],["1","0","1"]]))
test("islands: checkerboard 5 islands", r == 5, f"got {r}")
r = num_islands(grid([["1"]]))
test("islands: single cell", r == 1, f"got {r}")
r = num_islands(grid([["1","1"],["1","1"]]))
test("islands: all land 2x2", r == 1, f"got {r}")
r = num_islands(grid([["1","0"],["0","1"]]))
test("islands: diagonal not connected", r == 2, f"got {r}")

# can_finish
test("schedule: simple chain", can_finish(2, [[1, 0]]) is True)
test("schedule: two-node cycle", can_finish(2, [[1, 0], [0, 1]]) is False)
test("schedule: single course", can_finish(1, []) is True)
test("schedule: long chain", can_finish(5, [[1, 0], [2, 1], [3, 2], [4, 3]]) is True)
test("schedule: three-node cycle", can_finish(3, [[1, 0], [2, 1], [0, 2]]) is False)
test("schedule: diamond DAG", can_finish(4, [[1, 0], [2, 0], [3, 1], [3, 2]]) is True)
test("schedule: no prerequisites", can_finish(3, []) is True)

# word_ladder_length
r = word_ladder_length("hit", "cog", ["hot","dot","dog","lot","log","cog"])
test("ladder: standard (expect 5)", r == 5, f"got {r}")
r = word_ladder_length("hit", "cog", ["hot","dot","dog","lot","log"])
test("ladder: end_word not in list (expect 0)", r == 0, f"got {r}")
r = word_ladder_length("hot", "dog", ["hot","dog"])
test("ladder: direct one step (expect 2)", r == 2, f"got {r}")
r = word_ladder_length("a", "c", ["a","b","c"])
test("ladder: single char words (expect 2)", r == 2, f"got {r}")
r = word_ladder_length("abc", "xyz", ["ayz","axz","xyz"])
test("ladder: multi-step 3-char (expect 4)", r == 4, f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)

File: content/tracks/dsa-javascript/concepts/13-graphs-bfs-dfs/graphs-bfs-dfs_test.js
const assert = require("assert");
const { numIslands, canFinish, wordLadderLength } = require("./stub");

let passed = 0, failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e) {
    console.log(`  FAIL: ${name} — ${e.message}`);
    failed++;
  }
}

// numIslands
test("numIslands: single island", () => {
  assert.strictEqual(numIslands([
    ["1","1","1","1","0"],
    ["1","1","0","1","0"],
    ["1","1","0","0","0"],
    ["0","0","0","0","0"]
  ]), 1);
});
test("numIslands: three islands", () => {
  assert.strictEqual(numIslands([
    ["1","1","0","0","0"],
    ["1","1","0","0","0"],
    ["0","0","1","0","0"],
    ["0","0","0","1","1"]
  ]), 3);
});
test("numIslands: no islands", () => {
  assert.strictEqual(numIslands([["0","0"],["0","0"]]), 0);
});
test("numIslands: all land is one island", () => {
  assert.strictEqual(numIslands([["1","1"],["1","1"]]), 1);
});
test("numIslands: diagonal cells not connected", () => {
  assert.strictEqual(numIslands([["1","0"],["0","1"]]), 2);
});
test("numIslands: single cell land", () => {
  assert.strictEqual(numIslands([["1"]]), 1);
});
test("numIslands: single cell water", () => {
  assert.strictEqual(numIslands([["0"]]), 0);
});

// canFinish
test("canFinish: simple linear prerequisite", () => {
  assert.strictEqual(canFinish(2, [[1, 0]]), true);
});
test("canFinish: two-node cycle", () => {
  assert.strictEqual(canFinish(2, [[1, 0], [0, 1]]), false);
});
test("canFinish: no prerequisites", () => {
  assert.strictEqual(canFinish(1, []), true);
});
test("canFinish: 4 courses DAG", () => {
  assert.strictEqual(canFinish(4, [[1,0],[2,0],[3,1],[3,2]]), true);
});
test("canFinish: 3-node cycle", () => {
  assert.strictEqual(canFinish(3, [[0,1],[1,2],[2,0]]), false);
});
test("canFinish: many isolated courses", () => {
  assert.strictEqual(canFinish(5, []), true);
});
test("canFinish: chain 0->1->2->3", () => {
  assert.strictEqual(canFinish(4, [[1,0],[2,1],[3,2]]), true);
});

// wordLadderLength
test("wordLadderLength: classic hit->cog", () => {
  assert.strictEqual(
    wordLadderLength("hit", "cog", ["hot","dot","dog","lot","log","cog"]), 5
  );
});
test("wordLadderLength: endWord not in wordList", () => {
  assert.strictEqual(
    wordLadderLength("hit", "cog", ["hot","dot","dog","lot","log"]), 0
  );
});
test("wordLadderLength: direct one-step transform", () => {
  assert.strictEqual(wordLadderLength("a", "b", ["b"]), 2);
});
test("wordLadderLength: no path", () => {
  assert.strictEqual(wordLadderLength("hit", "cog", ["cog"]), 0);
});
test("wordLadderLength: longer path", () => {
  assert.strictEqual(
    wordLadderLength("hot", "dog", ["hot","dog","dot"]), 3
  );
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
